### PR TITLE
Expose and test new string.Join overloads

### LINKED ISSUE #5552: Add string.Join(char, ...) overloads
Moved from https://github.com/dotnet/coreclr/issues/2736#issuecomment-173295835 on behalf of @jasonwilliams200OK 

Current situation:

``` c#
char separator = ':';
string joined = string.Join(separator.ToString(), "one", "two", "three");
```

With `char` overload, users will be able to join string without calling `ToString()`
i.e. `string joined = string.Join(separator, "one", "two", "three");`

This will reduce at least one string allocation per call: https://github.com/dotnet/coreclr/blob/80c2952/src/mscorlib/src/System/Char.cs#L145-L168.
# 

Proposed API surface - @jasonwilliams200OK 

``` csharp
// in -> coreclr/src/mscorlib/src/System/String.cs
public static string Join(char separator, params object[] value)
public static string Join<T>(char separator, IEnumerable<T> values)
public static string Join(char separator, params string[] value)
public static string Join(char separator, string[] value, int startIndex, int count)

// REJECTED:
//public static string Join(char separator, IEnumerable<string> values)

// in -> coreclr/src/mscorlib/src/System/UnSafeCharBuffer.cs
public void AppendString(char charToAppend)
```

### PR DESCRIPTION
**Pull request by hughbe**
Fixes #5552

### CODE REVIEW

File: src/System.Runtime/tests/System/StringTests.netcoreapp.cs [hunk 1]
```diff
@@ -39,5 +40,82 @@ public static void StartsWith(string s, char value, bool expected)
...
+        {
+        	yield return new object[] { '|', new object[0], "" };
+        	yield return new object[] { '|', new object[] { 1 }, "1" };
+        	yield return new object[] { '|', new object[] { 1, 2, 3 }, "1|2|3" };
+        	yield return new object[] { '|', new object[] { new ObjectWithNullToString(), 2, new ObjectWithNullToString() }, "|2|" };
+        	yield return new object[] { '|', new object[] { "1", null, "3" }, "1||3" };
+        	yield return new object[] { '|', new object[] { "", "", "" }, "||" };
+        	yield return new object[] { '|', new object[] { "", null, "" }, "||" };
+        	yield return new object[] { '|', new object[] { null, null, null }, "||" };
+        }
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: Nit: indentation of the body

File: src/System.Runtime/tests/System/StringTests.netcoreapp.cs [hunk 2]
```diff
@@ -39,5 +40,82 @@ public static void StartsWith(string s, char value, bool expected)
...
+            Assert.Throws<ArgumentOutOfRangeException>("count", () => string.Join('|', new string[] { "Foo" }, 0, -1));
+        }
+
+        [Fact]
+        public static void Join_Char_InvalidStartIndexOut_ThrowsArgumentOutOfRangeException()
+        {
+            Assert.Throws<ArgumentOutOfRangeException>("startIndex", () => string.Join('|', new string[] { "Foo" }, 2, 1));
+            Assert.Throws<ArgumentOutOfRangeException>("startIndex", () => string.Join('|', new string[] { "Foo" }, 1, 2));
+            Assert.Throws<ArgumentOutOfRangeException>("startIndex", () => string.Join('|', new string[] { "Foo" }, 0, 2));
+        }
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: Can this be a theory?

File: src/System.Runtime/tests/System/StringTests.netcoreapp.cs [hunk 3]
```diff
@@ -39,5 +40,82 @@ public static void StartsWith(string s, char value, bool expected)
...
+
+        [Fact]
+        public static void Join_Char_NullValues_ThrowsArgumentNullException()
+        {
+        	Assert.Throws<ArgumentNullException>("value", () => string.Join('|', (string[])null));
+        	Assert.Throws<ArgumentNullException>("value", () => string.Join('|', (string[])null, 0, 0));
+        	Assert.Throws<ArgumentNullException>("values", () => string.Join('|', (object[])null));
+        	Assert.Throws<ArgumentNullException>("values", () => string.Join('|', (IEnumerable<object>)null));
+        }
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: nit: indentation

File: src/System.Runtime/tests/System/StringTests.netcoreapp.cs [hunk 4]
```diff
@@ -39,5 +40,82 @@ public static void StartsWith(string s, char value, bool expected)
...
+        {
+            if (startIndex == 0 && count == values.Length)
+            {
+                Assert.Equal(expected, string.Join(separator, values));
+                Assert.Equal(expected, string.Join(separator, (IEnumerable<string>)values));
+                Assert.Equal(expected, string.Join(separator, (object[])values));
+                Assert.Equal(expected, string.Join(separator, (IEnumerable<object>)values));
+            }
+
+            Assert.Equal(expected, string.Join(separator, values, startIndex, count));
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: To help verify that the behavior is the same with the string overloads, how about adding:
```C#
Assert.Equal(expected, string.Join(separator.ToString(), values, startindex, count));
```

File: src/System.Runtime/tests/System/StringTests.cs [hunk 5]
```diff
@@ -1607,19 +1607,31 @@ public static void Join_StringArray(string separator, string[] values, int start
...
+            Assert.Throws<ArgumentOutOfRangeException>("startIndex", () => string.Join("$$", new string[] { "Foo" }, -1, 0));
+        }
 
-            Assert.Throws<ArgumentOutOfRangeException>("startIndex", () => string.Join("$$", new string[] { "Foo" }, -1, 0)); // Start index < 0
-            Assert.Throws<ArgumentOutOfRangeException>("count", () => string.Join("$$", new string[] { "Foo" }, 0, -1)); // Count < 0
+        [Fact]
+        public static void Join_String_NegativeCount_ThrowsArgumentOutOfRangeException()
+        {
+            Assert.Throws<ArgumentOutOfRangeException>("count", () => string.Join("$$", new string[] { "Foo" }, 0, -1));
+        }
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.cs]: How about combining these two tests you split out into a theory?

File: src/System.Runtime/tests/System/StringTests.netcoreapp.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: Ahhh fixed

Comment (hughbe) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: Ahhh fixed

File: src/System.Runtime/tests/System/StringTests.netcoreapp.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: done

File: src/System.Runtime/tests/System/StringTests.netcoreapp.cs (hunk 3 above)
Comment (hughbe) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: fixed

File: src/System.Runtime/tests/System/StringTests.netcoreapp.cs (hunk 4 above)
Comment (hughbe) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: good idea, done

File: src/System.Runtime/tests/System/StringTests.cs (hunk 5 above)
Comment (hughbe) [inline src/System.Runtime/tests/System/StringTests.cs]: I've combined all the startIndex tests into a single theory test.

I kept the negaitve "count" tests on it's own, as the param name is different. Can change though if you want

**Comment by karelz** [comment]
Thanks @hughbe!

**Pull request merged**